bindings/sundials4py/cvodes/cvodes.cpp

```diff
@@ -0,0 +1,345 @@
+      fn_table->MEMBER = nb::cast(fn);                                            \
+      if (fn) { return NAME(cv_mem, which, &WRAPPER); }                           \
+      else { return NAME(cv_mem, which, nullptr); }                               \
+    },                                                                            \
+    __VA_ARGS__)
+
+#define BIND_CVODEB_CALLBACK2(NAME, FN_TYPE1, MEMBER1, WRAPPER1, FN_TYPE2, \
+                              MEMBER2, WRAPPER2, ...)                      \
+  m.def(                                                                   \
+    #NAME,                                                                 \
+    [](void* cv_mem, int which,                                            \
+       std::function<std::remove_pointer_t<FN_TYPE1>> fn1,                 \
+       std::function<std::remove_pointer_t<FN_TYPE2>> fn2)                 \
+    {                                                                      \
+      void* user_data   = nullptr;                                         \
+      auto fn_table     = get_cvodea_fn_table(cv_mem, which);              \
+      fn_table->MEMBER1 = nb::cast(fn1);                                   \
+      fn_table->MEMBER2 = nb::cast(fn2);                                   \
+      if (fn1) { return NAME(cv_mem, which, WRAPPER1, WRAPPER2); }         \
+      else { return NAME(cv_mem, which, nullptr, WRAPPER2); }              \
+    },                                                                     \
+    __VA_ARGS__)
+
+void bind_cvodes(nb::module_& m)
+{
+#include "cvodes_generated.hpp"
+
+  nb::class_<CVodeView>(m, "CVodeView")
+    .def("get", nb::overload_cast<>(&CVodeView::get, nb::const_),
+         nb::rv_policy::reference);
+
+  m.def(
+    "CVodeSetOptions",
+    [](void* cv_mem, const std::string& cvid, const std::string& file_name,
+       int argc, const std::vector<std::string>& args)
+    {
+      std::vector<char*> argv;
+
+      for (const auto& arg : args)
+      {
+        // We need a non-const char*, so we use data() and an explicit cast.
+        // This is safe as long as the underlying std::string is not modified.
+        argv.push_back(const_cast<char*>(arg.data()));
+      }
+
+      return CVodeSetOptions(cv_mem, cvid.empty() ? nullptr : cvid.c_str(),
+                             file_name.empty() ? nullptr : file_name.c_str(),
+                             argc, argv.data());
+    },
+    nb::arg("cv_mem"), nb::arg("cvid"), nb::arg("file_name"), nb::arg("argc"),
+    nb::arg("args"));
+
+  m.def(
+    "CVodeCreate",
+    [](int lmm, SUNContext sunctx)
+    { return std::make_shared<CVodeView>(CVodeCreate(lmm, sunctx)); },
+    nb::arg("lmm"), nb::arg("sunctx"), nb::keep_alive<0, 2>());
+
+  m.def("CVodeInit",
+        [](void* cv_mem, std::function<std::remove_pointer_t<CVRhsFn>> rhs,
+           sunrealtype t0, N_Vector y0)
+        {
+          int cv_status = CVodeInit(cv_mem, cvode_f_wrapper, t0, y0);
```

Comment:
Since we always pass `cvode_f_wrapper` to `CVodeInit` check if the rhs is `NULL` and check return flag
```suggestion
          if (!rhs) { throw sundials4py::illegal_value("rhs was null"); }

          int cv_status = CVodeInit(cv_mem, cvode_f_wrapper, t0, y0);
          if (cv_status != CV_SUCCESS)
          {
            throw sundials4py::error_returned("Failed to initialize CVODE");
          }
```
```

Reply:
Should we throw here or return the error code?

Reply:
I think we should return the error code instead (and throw only for the !rhs case)